### ISSUE TITLE
[JENKINS-23917] Protocol deadlock while uploading artifacts from ppc64

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#11939]
<p>I've encountered an ssh2 channel protocol issue when a ppc64 slave communicates with an x64 master.</p>

<p>Most operations, like sending build logs, work fine. When the time comes to upload artifacts at the end of the build the build stalls  indefinitely at:</p>

<pre>
Archiving artifacts
</pre>

<p>If I get stack dumps of slave and master using jstack, I see the master waiting to read from the slave:</p>

<pre>
<span class="code-quote">"Channel reader thread: Fedora16-ppc64-Power7-osuosl-karman"</span> prio=10 tid=0x00000000038c2800 nid=0x6de7 in <span class="code-object">Object</span>.wait() [0x00007f825ef8b000]
   java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
at java.lang.<span class="code-object">Object</span>.wait(Native Method)
- waiting on <0x00000000bf5802e0> (a com.trilead.ssh2.channel.Channel)
at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:502)
at com.trilead.ssh2.channel.FifoBuffer.read(FifoBuffer.java:212)
- locked <0x00000000bf5802e0> (a com.trilead.ssh2.channel.Channel)
at com.trilead.ssh2.channel.Channel$Output.read(Channel.java:127)
at com.trilead.ssh2.channel.ChannelManager.getChannelData(ChannelManager.java:946)
- locked <0x00000000bf5802e0> (a com.trilead.ssh2.channel.Channel)
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:58)
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:79)
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:82)
at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:67)
at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:93)
at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:33)
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)
</pre>

<p>and the slave is waiting for data from the master:</p>

<pre>
<span class="code-quote">"Channel reader thread: channel"</span> prio=10 tid=0x00000fff940fedd0 nid=0x558e runnable [0x00000fff6dc6d000]
   java.lang.<span class="code-object">Thread</span>.State: RUNNABLE
at java.io.FileInputStream.readBytes(Native Method)
at java.io.FileInputStream.read(FileInputStream.java:236)
at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
at java.io.BufferedInputStream.read(BufferedInputStream.java:254)
- locked <0x00000fff78ba9f98> (a java.io.BufferedInputStream)
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:82)
at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:67)
at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:93)
at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:33)
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)
</pre>

<p>of course I can't get those dumps at exactly the same moment, even if that were meaningful with network latencies and buffering, but repeated runs never show any other state for either thread.</p>

<p>tshark shows that there's some SSH chatter going on:</p>

<pre>
  0.000000 SLAVE -> MASTER SSH 126 Encrypted response packet len=60
  0.176121 MASTER -> SLAVE SSH 94 Encrypted request packet len=28
  0.176151 SLAVE -> MASTER TCP 66 ssh > 37501 [ACK] Seq=61 Ack=29 Win=707 Len=0 TSval=4141397874 TSecr=2808266826
</pre>

<p>but it should well be low level ssh keepalives or similar, as it's at precise 5 second intervals with nothing much else happening. There are three master->slave ssh connections, so it's not guaranteed that it's even the one associated with the stuck channel.</p>

<p>My first thought is endianness.</p>

<p>I don't really know how to begin debugging this issue, though.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ringerc.jpg" title="ringerc's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ringerc">ringerc</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-23917" target="_blank">Protocol deadlock while uploading artifacts from ppc64</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: core, remoting
<li><b>label(s)</b>: remoting, slave, ssh, unix
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2023-02-03T20:20:49+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I've encountered an ssh2 channel protocol issue when a ppc64 slave communicates with an x64 master.</p>

<p>Most operations, like sending build logs, work fine. When the time comes to upload artifacts at the end of the build the build stalls  indefinitely at:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Archiving artifacts
</pre>
</div></div>

<p>If I get stack dumps of slave and master using jstack, I see the master waiting to read from the slave:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"Channel reader thread: Fedora16-ppc64-Power7-osuosl-karman"</span> prio=10 tid=0x00000000038c2800 nid=0x6de7 in <span class="code-object">Object</span>.wait() [0x00007f825ef8b000]
   java.lang.<span class="code-object">Thread</span>.State: WAITING (on object monitor)
        at java.lang.<span class="code-object">Object</span>.wait(Native Method)
        - waiting on &lt;0x00000000bf5802e0&gt; (a com.trilead.ssh2.channel.Channel)
        at java.lang.<span class="code-object">Object</span>.wait(<span class="code-object">Object</span>.java:502)
        at com.trilead.ssh2.channel.FifoBuffer.read(FifoBuffer.java:212)
        - locked &lt;0x00000000bf5802e0&gt; (a com.trilead.ssh2.channel.Channel)
        at com.trilead.ssh2.channel.Channel$Output.read(Channel.java:127)
        at com.trilead.ssh2.channel.ChannelManager.getChannelData(ChannelManager.java:946)
        - locked &lt;0x00000000bf5802e0&gt; (a com.trilead.ssh2.channel.Channel)
        at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:58)
        at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:79)
        at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:82)
        at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:67)
        at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:93)
        at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:33)
        at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
        at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)
</pre>
</div></div>

<p>and the slave is waiting for data from the master:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-quote">"Channel reader thread: channel"</span> prio=10 tid=0x00000fff940fedd0 nid=0x558e runnable [0x00000fff6dc6d000]
   java.lang.<span class="code-object">Thread</span>.State: RUNNABLE
        at java.io.FileInputStream.readBytes(Native Method)
        at java.io.FileInputStream.read(FileInputStream.java:236)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:254)
        - locked &lt;0x00000fff78ba9f98&gt; (a java.io.BufferedInputStream)
        at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:82)
        at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:67)
        at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:93)
        at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:33)
        at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
        at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)
</pre>
</div></div>

<p>of course I can't get those dumps at exactly the same moment, even if that were meaningful with network latencies and buffering, but repeated runs never show any other state for either thread.</p>

<p>tshark shows that there's some SSH chatter going on:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">  0.000000 SLAVE -&gt; MASTER SSH 126 Encrypted response packet len=60
  0.176121 MASTER -&gt; SLAVE SSH 94 Encrypted request packet len=28
  0.176151 SLAVE -&gt; MASTER TCP 66 ssh &gt; 37501 [ACK] Seq=61 Ack=29 Win=707 Len=0 TSval=4141397874 TSecr=2808266826
</pre>
</div></div>

<p>but it should well be low level ssh keepalives or similar, as it's at precise 5 second intervals with nothing much else happening. There are three master-&gt;slave ssh connections, so it's not guaranteed that it's even the one associated with the stuck channel.</p>

<p>My first thought is endianness.</p>

<p>I don't really know how to begin debugging this issue, though.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins master ver. 1.574-SNAPSHOT (git 17d9093) on x64 Debian 7 with:<br/>
java version &quot;1.6.0_32&quot;<br/>
OpenJDK Runtime Environment (IcedTea6 1.13.4) (6b32-1.13.4-1~deb7u1)<br/>
OpenJDK 64-Bit Server VM (build 23.25-b01, mixed mode)<br/>
Jenkins slave on POWER7 ppc64 Fedora 16 (an OSU OSL community machine), with Java:<br/>
java version &quot;1.6.0_24&quot;<br/>
OpenJDK Runtime Environment (IcedTea6 1.11.1) (fedora-65.1.11.1.fc16-ppc64)<br/>
OpenJDK 64-Bit Zero VM (build 20.0-b12, interpreted mode)
```
</details>
<details><summary><i>6 attachments</i></summary>

- [config.xml](https://issues.jenkins.io/secure/attachment/26420/config.xml)
- [jenkins-master-idle-stack.txt](https://issues.jenkins.io/secure/attachment/26423/jenkins-master-idle-stack.txt)
- [jenkins-master-stack.txt](https://issues.jenkins.io/secure/attachment/26421/jenkins-master-stack.txt)
- [jenkins-slave-stack.txt](https://issues.jenkins.io/secure/attachment/26422/jenkins-slave-stack.txt)
- [slavelog-from-master.txt](https://issues.jenkins.io/secure/attachment/26424/slavelog-from-master.txt)
- [slavelog-from-slave.txt](https://issues.jenkins.io/secure/attachment/26425/slavelog-from-slave.txt)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ringerc.jpg" title="ringerc's avatar" /> ringerc's <a href="https://issues.jenkins.io/browse/JENKINS-23917?focusedId=206223&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-206223">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I'm going to re-test with a simplified build and a single executor configured.</p>
</pre>
</details>

</details>
<p>I'm going to re-test with a simplified build and a single executor configured.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ringerc.jpg" title="ringerc's avatar" /> ringerc's <a href="https://issues.jenkins.io/browse/JENKINS-23917?focusedId=206296&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-206296">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>When I switch the node to JNLP (so it uses direct TCP/IP as a transport, rather than SSH) I can no longer reproduce this despite repeated test runs.</p>

<p>So, so far:</p>

<ul class="alternate" type="square">
	<li>Only observed on ppc64</li>
	<li>Only observed for ssh slaves</li>
	<li>Using -text protocol does not help</li>
</ul>

</pre>
</details>

</details>
<p>When I switch the node to JNLP (so it uses direct TCP/IP as a transport, rather than SSH) I can no longer reproduce this despite repeated test runs.</p>

<p>So, so far:</p>

<ul class="alternate" type="square">
	<li>Only observed on ppc64</li>
	<li>Only observed for ssh slaves</li>
	<li>Using -text protocol does not help</li>
</ul>
